util/network/cniprovider/createns_linux.go

```diff
@@ -38,63 +38,51 @@ func cleanOldNamespaces(c *cniProvider) {
 	}()
 }
 
-func createNetNS(c *cniProvider, id string) (string, error) {
+// unshareAndMount needs to be called in a separate thread
+func unshareAndMountNetNS(target string) error {
+	if err := syscall.Unshare(syscall.CLONE_NEWNET); err != nil {
+		return err
+	}
+
+	return syscall.Mount("/proc/thread-self/ns/net", target, "", syscall.MS_BIND, "")
+}
+
+func createNetNS(c *cniProvider, id string) (_ string, err error) {
 	nsPath := filepath.Join(c.root, "net/cni", id)
 	if err := os.MkdirAll(filepath.Dir(nsPath), 0700); err != nil {
 		return "", err
 	}
 
 	f, err := os.Create(nsPath)
 	if err != nil {
-		deleteNetNS(nsPath)
 		return "", err
 	}
+	defer func() {
+		if err != nil {
+			deleteNetNS(nsPath)
+		}
+	}()
 	if err := f.Close(); err != nil {
-		deleteNetNS(nsPath)
-		return "", err
-	}
-	procNetNSBytes, err := syscall.BytePtrFromString("/proc/self/ns/net")
-	if err != nil {
-		deleteNetNS(nsPath)
 		return "", err
 	}
-	nsPathBytes, err := syscall.BytePtrFromString(nsPath)
-	if err != nil {
-		deleteNetNS(nsPath)
-		return "", err
-	}
-	beforeFork()
 
-	pid, _, errno := syscall.RawSyscall6(syscall.SYS_CLONE, uintptr(syscall.SIGCHLD)|unix.CLONE_NEWNET, 0, 0, 0, 0, 0)
-	if errno != 0 {
-		afterFork()
-		deleteNetNS(nsPath)
-		return "", errno
-	}
+	errCh := make(chan error)
 
-	if pid != 0 {
-		afterFork()
-		var ws unix.WaitStatus
-		_, err = unix.Wait4(int(pid), &ws, 0, nil)
-		for err == syscall.EINTR {
-			_, err = unix.Wait4(int(pid), &ws, 0, nil)
-		}
+	go func() {
+		defer close(errCh)
+		runtime.LockOSThread()
 
-		if err != nil {
-			deleteNetNS(nsPath)
-			return "", errors.Wrapf(err, "failed to find pid=%d process", pid)
+		if err := unshareAndMountNetNS(nsPath); err != nil {
+			errCh <- err
 		}
-		errno = syscall.Errno(ws.ExitStatus())
-		if errno != 0 {
-			deleteNetNS(nsPath)
-			return "", errors.Wrapf(errno, "failed to mount %s (pid=%d)", nsPath, pid)
-		}
-		return nsPath, nil
+
+		// we leave the thread locked so go runtime terminates the thread
```

Comment:
The effects of `unshare(CLONE_NEWNET)` are fully reversible. The thread can be unlocked (returned to the runtime) if the thread's network namespace is first reverted using `setns`. Package [`github.com/docker/docker/internal/unshare`](https://github.com/moby/moby/blob/d7e6e33d07a5bafa7a410bd66e22c64f2c656360/internal/unshare/unshare_linux.go) abstracts away those details; maybe it's time we promote that package to a `moby/sys` module so it can be imported here instead of reinventing the wheel.

Reply:
I tend to think it is not worth it to do the reversing. The performance impact should be minimal but internal logic would be much more complicated.

Maybe if we used a shared library, then it would probably be fine. But there are some other differences, eg. I'm not sure if locking `init()` like it is done in that library makes sense if you don't unshare the mount namespace.

Reply:
> not sure if locking init()

I guess it wouldn't matter as when docker is updated, buildkit will get that behavior anyway as we import `chrootarchive`.

Reply:
Could also use [gonso](https://github.com/cpuguy83/gonso) for this.